### PR TITLE
Add option for incremental evaluation tests for quantifier instantiation (ieval)

### PR DESCRIPTION
**Pull request by ajreynol**
This adds an option `ieval` for incremental instantiation evaluation tests. These tests are run as part of the implementation `InstMatch` objects (currently being used by E-matching). When the `ieval` option is enabled, we consult the appropriate instantiation evaluator to check if the match is feasible.

It furthermore extends conflict-based instantiation to maintain `InstMatch` objects, hence allowing entailment to be tested eagerly and incrementally in conflict-based instantiation.

I am keeping this disabled by default still for now, but it will likely be enabled by default soon after further testing.

`--ieval=use` is +209-132 on SMT-LIB quantified logics compared with default. `--ieval=use-learn` is slightly worse than `--ieval=use` (-14+8).
```
                           Configuration  #unsat      time    #sat      time #solved    #total
                cvc5-092122-ieval_iev_ul   63779   26064.4    1529   476.598     65308     91390
                   cvc5-092122-ieval_iev   63785   25976.3    1529   477.706     65314     91390
                     cvc5-092122-ieval_d   63708   25613.9    1529   477.386     65237     91390




For unsatisfiable : 
                     cvc5-092122-ieval_d       X     132     132
                   cvc5-092122-ieval_iev     209       X      14
                cvc5-092122-ieval_iev_ul     203       8       X

For satisfiable : 
                     cvc5-092122-ieval_d       X       0       0
                   cvc5-092122-ieval_iev       0       X       0
                cvc5-092122-ieval_iev_ul       0       0       X
```

Adds a few regressions where the old and new entailment tests differed in my debugging (now fixed).

### CODE REVIEW

File: src/theory/quantifiers/quant_conflict_find.cpp [hunk 1]
```diff
@@ -610,27 +621,52 @@ bool QuantInfo::setMatch(size_t v, TNode n, bool isGroundRep, bool isGround)
...
+    d_vars_set.insert(v);
+    Trace("qcf-match-debug")
+        << "---- now bound " << d_vars_set.size() << " / "
+        << d_q[0].getNumChildren() << " base variables." << std::endl;
+  }
+  // Note that assigning to a variable that an original variable is equal to
+  // should trigger the match object. For example, if we have auxiliary
+  // variable k and original variable x where x <-> k currently, and we set
+  // k -> t, then we should notify the match object that x -> t. However,
+  // this is not done here, as it would require more complex bookkeeping.
```
Comment (HanielB) [inline src/theory/quantifiers/quant_conflict_find.cpp:648]: clarify if it's done somewhere else. Also clarify the implications of not doing it if it's not done at all.

File: src/theory/quantifiers/quant_conflict_find.cpp [hunk 2]
```diff
@@ -649,6 +685,13 @@ bool QuantInfo::isMatchSpurious()
 
 bool QuantInfo::isTConstraintSpurious(const std::vector<Node>& terms)
 {
+  if (options().quantifiers.ievalMode != options::IevalMode::OFF)
+  {
+    // We rely on the instantiation evaluator. When the instantiation evaluator
+    // is enabled, this method (almost) always returns false. It may return
+    // true based on minor differences in the entailment tests.
```
Comment (HanielB) [inline src/theory/quantifiers/quant_conflict_find.cpp:692]: Clarify the consequences of this method returning `true` when the evaluator would have said `false`. Moreover what are the kind of differences that are at play here? Is it a matter of incompleteness?

Comment (ajreynol) [reply src/theory/quantifiers/quant_conflict_find.cpp]: I don't concretely know the difference, the same general entailment test is being run, although by 2 very different pieces of code.

File: src/theory/quantifiers/quant_conflict_find.cpp [hunk 3]
```diff
@@ -610,27 +621,54 @@ bool QuantInfo::setMatch(size_t v, TNode n, bool isGroundRep, bool isGround)
...
+        << "---- now bound " << d_vars_set.size() << " / "
+        << d_q[0].getNumChildren() << " base variables." << std::endl;
+  }
+  // Note that assigning to a variable that an original variable is equal to
+  // should trigger the match object. For example, if we have auxiliary
+  // variable k and original variable x where x <-> k currently, and we set
+  // k -> t, then we should notify the match object that x -> t. However,
+  // this is not done, as it would require more complex bookkeeping. Overall,
+  // this means that we may fail in some rare cases to eagerly recognize when a
+  // substitution is an entailed.
```
Comment (HanielB) [inline src/theory/quantifiers/quant_conflict_find.cpp:650]:
```suggestion
  // substitution is entailed.
```

**Review by HanielB — APPROVED** [review]
Just a minor point

**Pull request merged**